metadata-provider/src/contract.rs

```diff
@@ -0,0 +1,271 @@
+use cosmwasm_std::{
+    debug_print, to_binary, Api, CanonicalAddr, Env, Extern, HandleResponse, HandleResult,
+    HumanAddr, InitResponse, InitResult, Querier, QueryResult, ReadonlyStorage, StdError,
+    StdResult, Storage,
+};
+use cosmwasm_storage::{PrefixedStorage, ReadonlyPrefixedStorage};
+
+use crate::msg::{
+    HandleAnswer, HandleMsg, InitMsg, QueryAnswer, QueryMsg, ResponseStatus::Success,
+};
+use crate::state::{
+    load, may_load, save, Config, BLOCK_KEY, CONFIG_KEY, PREFIX_PRIV_META, PREFIX_PUB_META,
+    PREFIX_VIEW_KEY,
+};
+use crate::viewing_key::{ViewingKey, VIEWING_KEY_SIZE};
+
+use secret_toolkit::snip721::{Metadata, ViewerInfo};
+use secret_toolkit::utils::{pad_handle_result, pad_query_result};
+
+/// pad handle responses and log attributes to blocks of 256 bytes to prevent leaking info based on
+/// response size
+pub const BLOCK_SIZE: usize = 256;
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: InitMsg,
+) -> InitResult {
+    let config = Config {
+        name: msg.name,
+        symbol: msg.symbol,
+        admin: deps.api.canonical_address(&env.message.sender)?,
+        token_address: deps.api.canonical_address(&msg.token_address)?,
+        token_code_hash: msg.token_code_hash,
+    };
+
+    save(&mut deps.storage, CONFIG_KEY, &config)?;
+
+    debug_print!("Contract was initialized by {}", env.message.sender);
+
+    Ok(InitResponse::default())
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> HandleResult {
+    save(&mut deps.storage, BLOCK_KEY, &env.block)?;
+    let mut config: Config = load(&deps.storage, CONFIG_KEY)?;
```

Comment:
As a gas saving measure only load this inside the functions that need it, I've made the same mistake too a few times. 

SetMetadata doesn't actually use the config argument